tools/distribution/distrotest.py

```diff
@@ -0,0 +1,715 @@
+import logging
+import pathlib
+import re
+import shutil
+from functools import cached_property
+from itertools import chain
+from typing import Callable, Iterable, List, Optional, Tuple, Type, Union
+
+import verboselogs
+
+import aiodocker
+
+from tools.base import checker, utils
+from tools.docker import utils as docker_utils
+
+DISTROTEST_CONFIG_PATH = "tools/distribution/distrotest.yaml"
+
+DOCKER_IMAGE_PREFIX = "envoybuild_"
+DOCKER_CONTAINER_PREFIX = "envoytest_"
+DOCKERFILE_TEMPLATE = """
+FROM {build_image}
+{env}
+
+ADD {install_dir} {install_mount_path}
+ADD {testfile_name} {test_mount_path}
+ADD {keyfile_name} {key_mount_path}
+RUN {build_command}
+
+CMD ["tail", "-f", "/dev/null"]
+"""
+
+
+class BuildError(Exception):
+    pass
+
+
+class ConfigurationError(Exception):
+    pass
+
+
+class ContainerError(Exception):
+    pass
+
+
+class DistroTestConfig(object):
+    """Configuration object for distro tests
+
+    This holds the configuration (and Docker connect) across a batch of tests.
+
+    It also allows `DistroTest` to adapt `checker.Checker`.
+
+    Init parameters:
+
+    `path` is the path to the directory that will be used as the Docker context.
+    `tarball` is the path to a tarball containing packages.
+    `keyfile` is the path to the public key used to sign the packages
+    `testfile` is the bash script to run inside the test containers.
+    `maintainer` is the expected maintainer/packager the packages were signed with.
```

Comment:
i think its theoretically possible to derive the maintainer from the `keyfile` but probs i would prefer to follow up with this as it adds some complexity